lib/hdb/common.c

```diff
@@ -153,6 +154,23 @@ _hdb_fetch_kvno(krb5_context context, HDB *db, krb5_const_principal principal,
 	    krb5_data_free(&value);
 	    return ret;
 	}
+
+	if ((flags & HDB_F_FOR_AS_REQ) && (flags & HDB_F_CANON) == 0) {
+	    krb5_principal tmp;
+
+	    /*
+	     * Return the name the client asked for. Should be done in the KDC,
```

Comment:
I actually like this being done in `libhdb`, honestly.  We already have a bunch of KDC specials, such as if the caller requests a specific kvno then `libhdb` puts that in the `keys` field so the caller doesn't have to go looking through key history (OK, I did that), but that's good if you posit a very normalized backend schema and you don't want to have to pull in things the caller doesn't need.

Reply:
I'm with you 50/50. I kind of like it too. But, I also like the idea of normalizing the AS path so that it's identical to the TGS path: if you ask for a principal with name canonicalization disabled, the name in the reply is the one you asked for, and this is done by the KDC.